pyOCD/tools/pyocd.py

```diff
@@ -69,207 +76,490 @@ def dumpHexData(data, startAddress=0, width=8):
-        parser.add_argument("data", nargs='*', help="Data to write using the --write option")
-        args = parser.parse_args()
-
-        # Set read/write actions.
-        if args.read:
-            args.action = ACTION_READ
-            args.read = int(args.read, base=0)
-            args.len = int(args.len, base=0)
-        elif args.write:
-            args.action = ACTION_WRITE
-            args.write = int(args.write, base=0)
-            args.data = [int(d, base=0) for d in args.data]
-
-        # Default to list mode if no other action was specified.
-        if args.action == None and args.file == None:
-            args.action = ACTION_LIST
-        elif args.file:
-            args.action = ACTION_PROGRAM
-
-        return args
+        parser.add_argument("-d", "--debug", dest="debug_level", choices=debug_levels, default='warning', help="Set the level of system logging output. Supported choices are: "+", ".join(debug_levels), metavar="LEVEL")
+        parser.add_argument("cmd", nargs='?', default=None, help="Command")
+        parser.add_argument("args", nargs='*', help="Arguments for the command.")
+        return parser.parse_args()
 
-    def run(self):
-        board = None
-        exitCode = 0
+    def configure_logging(self):
+        level = LEVELS.get(self.args.debug_level, logging.WARNING)
+        logging.basicConfig(level=level)
 
+    def run(self):
         try:
             # Read command-line arguments.
-            args = self.get_args()
-
-            if args.verbose:
-                logging.basicConfig(level=logging.INFO)
-
-            # Print a list of all connected boards.
-            if args.action == ACTION_LIST:
-                MbedBoard.listConnectedBoards()
-                sys.exit(0)
-
-            board = MbedBoard.chooseBoard(board_id=args.board, target_override=args.target, init_board=False)
-            board.target.setAutoUnlock(False)
+            self.args = self.get_args()
+            self.cmd = self.args.cmd
+            if self.cmd:
+                self.cmd = self.cmd.lower()
+
+            # Set logging level
+            self.configure_logging()
+
+            # Check for a valid command.
+            if self.cmd and self.cmd not in self.command_list:
+                print "Error: unrecognized command '%s'" % self.cmd
+                return 1
+
+            # List command must be dealt with specially.
+            if self.cmd == 'list':
+                self.handle_list([])
+                return 0
+
+            if self.args.clock != DEFAULT_CLOCK_FREQ_KHZ:
+                print "Setting SWD clock to %d kHz" % self.args.clock
+
+            # Connect to board.
+            self.board = MbedBoard.chooseBoard(board_id=self.args.board, target_override=self.args.target, init_board=False, frequency=(self.args.clock * 1000))
+            self.board.target.setAutoUnlock(False)
+            self.board.target.setHaltOnConnect(False)
```

Comment:
A prototype for these functions should probably be added in target.py, along with a stub in cortex_m.py.  That way targets without this will be able to use pyocd tool.